contrib/drivers/ceph/replication.go

```diff
@@ -0,0 +1,407 @@
+// Copyright 2018 The OpenSDS Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations
+// under the License.
+
+package ceph
+
+import (
+	"github.com/astaxie/beego/logs"
+	log "github.com/golang/glog"
+	. "github.com/opensds/opensds/contrib/drivers/utils/config"
+	"github.com/opensds/opensds/pkg/model"
+	pb "github.com/opensds/opensds/pkg/model/proto"
+	"github.com/opensds/opensds/pkg/utils/config"
+	"golang.org/x/crypto/ssh"
+	"time"
+)
+
+type ReplicationDriver struct {
+	conf *Config
+}
+
+type Replication struct {
+	IPaddresshost string `yaml:"hostip,omitempty"`
+	Ipaddresspeer string `yaml:"peerip,omitempty"`
+	Username      string `yaml:"username,omitempty"`
+	Password      string `yaml:"password,omitempty"`
+	HostdailIP    string `yaml:"hostdailip,omitempty"`
+	PeerdailIP    string `yaml:"peerdailip,omitempty"`
+}
+
+type Config struct {
+	ConfigFile  string `yaml:"configFile,omitempty"`
+	Replication `yaml:"replication"`
+}
+
+// Setup
+func (r *ReplicationDriver) Setup() error {
+
+	r.conf = &Config{ConfigFile: "/etc/ceph/ceph.conf"}
+	p := config.CONF.OsdsDock.Backends.Ceph.ConfigPath
+	if "" == p {
+		p = defaultConfPath
+	}
+	_, err := Parse(r.conf, p)
+	if err != nil {
+		logs.Error(err)
+	}
+
+	return nil
+}
+
+// Unset
+func (r *ReplicationDriver) Unset() error { return nil }
+
+// Create replication
+
+func (r *ReplicationDriver) CreateReplication(opt *pb.CreateReplicationOpts) (*model.ReplicationSpec, error) {
+
+	volumename := opensdsPrefix + opt.PrimaryVolumeId
+
+	cephconfig := &ssh.ClientConfig{
+		User: r.conf.Username,
+		Auth: []ssh.AuthMethod{
+			ssh.Password(r.conf.Password),
+		},
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+	}
+	cephclient, err := ssh.Dial("tcp", r.conf.HostdailIP, cephconfig)
+	if err != nil {
+		log.Error("Failed to dial: " + err.Error())
+
+	}
+
+	backupconfig := &ssh.ClientConfig{
+		User: r.conf.Username,
+		Auth: []ssh.AuthMethod{
+			ssh.Password(r.conf.Password),
+		},
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+	}
+	backupclient, error := ssh.Dial("tcp", r.conf.PeerdailIP, backupconfig)
+
+	if error != nil {
+		log.Error("Failed to dial: " + error.Error())
+
+	}
+
+	cephenablesession, err := cephclient.NewSession()
+	if err != nil {
+	}
+
+	cephenablesession.Run("rbd  feature enable rbd/" + volumename + " exclusive-lock,journaling")
+	cephenablesession.Close()
+
+	cephrbdinstall, err := cephclient.NewSession()
+	if err := cephrbdinstall.Run("apt install -y rbd-mirror"); err != nil {
```

Comment:
what if i run replicate the 2nd time ? do we need to check - if rbd not installed, then install. else go on to replication